File: GVO Code v3/handpadle.pas
unit handpadle;

interface

uses utils, sysutils, windows, dialogs;

procedure HandleHandpadle;

implementation

uses Main, OMS68SERMC, ManualPadle, qgraphics;

procedure HandleHandpadle;
var
	CmdStr, TempStr, xspeed, yspeed: String;
	i: Integer;
Begin
	EastPushed := False; WestPushed := False; NorthPushed := False; SouthPushed := False;
	SendAndGetString(CommRecord, 'BX;', Response);
	Tempstr := Response;
	if (length(tempstr) <> 2) Then exit;
	//Handle Buttons
	io := Byte(TempStr[1]);                    {get First Char}
	if (io > $39) Then io := io - $37
	Else io := io - $30;
	//
	//manual bit is high speed slew
	//
	if (io AND ManualBit <> 0) Then manualPushed := True else ManualPushed := false;
	if (io AND SouthBit <> 0) Then SouthPushed := True else SouthPushed := false;
	//
	io := Byte(TempStr[2]);                    {get Last Char}
	if (io > $39) Then io := io - $37
	Else io := io - $30;
	if (io AND NorthBit <> 0) Then NorthPushed := True else NorthPushed := false;
	if (io AND EastBit <> 0) Then EastPushed := True else EastPushed := false;
	if (io AND WestBit <> 0) Then WestPushed := True else WestPushed := false;
	//
	If (EastPushed) Then MainForm.Label7.Color := clred else MainForm.Label7.Color := clsilver;
	If (WestPushed) Then MainForm.Label8.Color := clred else MainForm.Label8.Color := clsilver;
	If (NorthPushed) Then MainForm.label5.Color := clred else MainForm.label5.Color := clsilver;
	If (SouthPushed) Then MainForm.label6.Color := clred else MainForm.label6.Color := clsilver;
	//  If(manualPushed)Then MainForm.Color := clyellow else MainForm.Color := clBtnFace;

	//
	If ((ManualPushed) OR (MainForm.Select5Inch.Checked = True)) Then
	Begin
		HandleFastPadle;
		Exit;
	End
	Else
	Begin
		xspeed := xvl;
		yspeed := yvl;
	End;
	//
	// Hand padle no high speed
	//
	//  correcting is slow correct
	//
	//  Slewing is fast correcting
	//
	//
	{X Axis (RA) Update}
	Case Xstate of
		Off: Begin
			End;
		Tracking: Begin
			If (EastPushed) Then
			Begin
				Xstate := CorrectingE;
				i := Round(TrkRate / 2);
				Str(i, TempStr);
				CmdStr := 'AX JG' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));          //slow track by 1/2
			End;
			{}
			If (WestPushed) Then
			Begin
				Xstate := CorrectingW;
				i := Round(TrkRate + (TrkRate / 2));
				Str(i, TempStr);
				CmdStr := 'AX JG' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));        //speed up track by 1/2
			End;
			{}
		End;
		CorrectingE: Begin
			If ((NOT EastPushed) AND (NOT WestPushed)) Then
			Begin                        //go to track
				Xstate := Tracking;
				CmdStr := 'AX ST ;';
				SendString(CommRecord, pchar(CmdStr)); SendString(CommRecord, pchar(CmdStr));
				Str(TrkRate:10:6, TempStr);
				CmdStr := 'AX JF' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
			If ((EastPushed) AND (WestPushed)) Then
			Begin       //start intermediate slew  to east
				XState := Slewing;
				CmdStr := 'AX ST; AX JG-' + XSpeed + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
		CorrectingW: Begin
			If ((NOT EastPushed) AND (NOT WestPushed)) Then
			Begin                         //start tracking
				Xstate := Tracking;
				CmdStr := 'AX ST ;';
				SendString(CommRecord, pchar(CmdStr)); SendString(CommRecord, pchar(CmdStr));
				Str(TrkRate:10:6, TempStr);
				CmdStr := 'AX JF' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
			If ((EastPushed) AND (WestPushed)) Then
			Begin                 //start intermediate slew to west
				XState := Slewing;
				CmdStr := 'AX JG' + XSpeed + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
			{}
		End;
		Slewing: Begin
			If ((NOT EastPushed) AND (NOT WestPushed)) Then
			Begin
				XState := Tracking;
				Str(TrkRate:10:6, TempStr);
				CmdStr := 'AX ST;';
				SendString(CommRecord, pchar(CmdStr));
				CmdStr := 'AX JF' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
	Else;
	End; {X State case}
	{}
	{Y Axis (dec) Update}
	//
	// bump north or south for future ant- backlas
	//
	Case Ystate of
		Off: Begin
			End;
		Tracking: Begin
			If (NorthPushed) Then
			Begin
				if ((Ypole >= 0) AND (LastDecNorth)) Then
					BumpSouth
				else
					If (NOT LastDecNorth) Then BumpNorth;
				Ystate := CorrectingN;
				i := Round(TrkRate / 4);
				Str(i, TempStr);
				CmdStr := ' AY JG' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
			{}
			If (SouthPushed) Then
			Begin
				if ((Ypole >= 0) AND (Not LastDecNorth)) Then
					BumpNorth
				else
					If (LastDecNorth) Then BumpSouth;
				Ystate := CorrectingS;
				i := Round(TrkRate / 4);
				Str(i, TempStr);
				CmdStr := ' AY JG-' + TempStr + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
		//
		// end of tracking state
		//
		CorrectingN: Begin
			If ((NOT NorthPushed) AND (NOT SouthPushed)) Then
			Begin
				Ystate := Tracking;
				CmdStr := ' AY ST;';
				SendString(CommRecord, pchar(CmdStr));
			End;
			If ((NorthPushed) AND (SouthPushed)) Then
			Begin
				Ystate := Slewing;
				CmdStr := ' AY JG' + ySpeed + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
		CorrectingS: Begin
			If ((NOT NorthPushed) AND (NOT SouthPushed)) Then
			Begin
				Ystate := Tracking;
				CmdStr := ' AY ST;';
				SendString(CommRecord, pchar(CmdStr));
			End;
			If ((NorthPushed) AND (SouthPushed)) Then
			Begin
				Ystate := Slewing;
				CmdStr := ' AY JG-' + ySpeed + ';';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
		{}
		Slewing: Begin
			If ((NOT NorthPushed) AND (NOT SouthPushed)) Then
			Begin
				Ystate := Tracking;
				CmdStr := ' AY ST; ';
				SendString(CommRecord, pchar(CmdStr));
			End;
		End;
	End; {Y State Case}
End;

end.

File: GVO Code C++ v1/HandPadle.cpp
#include "HandPadle.h"
#include "CommUtils.h"
#include "OMS68SERMC.h"
#include "CalcCoord.h"
#include "TimerUnit.h"
#include "StateVar.h"
#include "utils.h"
#include "GlobalValues.h"
#include "ManualPadle.h"
#include "Bits.h"
#include <string>
#include <cmath>
#include <iostream>
#include <cctype>
#include <cstdint>

void HandleHandPadle() {
	if (movingRA || movingDEC) {
		return;
	}

	std::string CmdStr, TempStr;
	int i;

	static bool prevNorth = false;
	static bool prevSouth = false;
	static bool prevEast  = false;
	static bool prevWest  = false;

	// Reset button states
	EastPushed = false;
	WestPushed = false;
	NorthPushed = false;
	SouthPushed = false;
	ManualPushed = false;

	// Read handpad
	SendAndGetCommand(&CommRecord, "BX;", Response, sizeof(Response));
	TempStr = Response;

	if (TempStr.length() != 2) {
		return;
	}

	// Decode first nibble
	uint8_t ioLocal = std::toupper(static_cast<unsigned char>(TempStr[0]));
	ioLocal = (ioLocal > '9') ? ioLocal - 0x37 : ioLocal - 0x30;

	ManualPushed = (ioLocal & ManualBit) != 0;
	SouthPushed  = (ioLocal & SouthBit)  != 0;

	// Decode second nibble
	ioLocal = std::toupper(static_cast<unsigned char>(TempStr[1]));
	ioLocal = (ioLocal > '9') ? ioLocal - 0x37 : ioLocal - 0x30;

	NorthPushed = (ioLocal & NorthBit) != 0;
	EastPushed  = (ioLocal & EastBit)  != 0;
	WestPushed  = (ioLocal & WestBit)  != 0;

	// Print ONLY on press edge
	if ((EastPushed && !prevEast) ||
		(WestPushed && !prevWest) ||
		(NorthPushed && !prevNorth) ||
		(SouthPushed && !prevSouth)) {

		// std::cout << "[HandPad] Press -> " << "E:" << EastPushed << " W:" << WestPushed << " N:" << NorthPushed << " S:" << SouthPushed << "\n";
	}

	prevEast  = EastPushed;
	prevWest  = WestPushed;
	prevNorth = NorthPushed;
	prevSouth = SouthPushed;

	// Fast mode (button held)
	if (ManualPushed) {
		HandleFastPadle();
		return;
	}

	switch (Xstate) {
			case StateVar::Tracking:
				if (EastPushed) {
					Xstate = StateVar::CorrectingE;

					i = (int)std::round(TrkRate / 2.0);

					CmdStr = "AX JG" + std::to_string(i) + ";";
					SendCommand(CmdStr);
				}

				if (WestPushed) {
					Xstate = StateVar::CorrectingW;

					i = (int)std::round(TrkRate + TrkRate / 2.0);

					CmdStr = "AX JG" + std::to_string(i) + ";";
					SendCommand(CmdStr);
				}
				break;

			case StateVar::CorrectingE:
				if (!EastPushed && !WestPushed) {
					Xstate = StateVar::Tracking;

					SendCommand("AX ST;");
					SendCommand("AX ST;");

					CmdStr = "AX JF" + FormatDouble(TrkRate, 10, 6) + ";";
					SendCommand(CmdStr);
				}

				if (EastPushed && WestPushed) {
					Xstate = StateVar::Slewing;

					CmdStr = "AX ST; AX JG-" + xvl + ";";
					SendCommand(CmdStr);
				}
				break;

			case StateVar::CorrectingW:
				if (!EastPushed && !WestPushed) {
					Xstate = StateVar::Tracking;

					SendCommand("AX ST;");
					SendCommand("AX ST;");

					CmdStr = "AX JF" + FormatDouble(TrkRate, 10, 6) + ";";
					SendCommand(CmdStr);
				}

				if (EastPushed && WestPushed) {
					Xstate = StateVar::Slewing;

					CmdStr = "AX JG" + xvl + ";";
					SendCommand(CmdStr);
				}
				break;

			case StateVar::Slewing:
				if (!EastPushed && !WestPushed) {
					Xstate = StateVar::Tracking;

					SendCommand("AX ST;");

					CmdStr = "AX JF" + FormatDouble(TrkRate, 10, 6) + ";";
					SendCommand(CmdStr);
				}
				break;

			default:
				break;
			}

	switch (Ystate) {
	case StateVar::Tracking:
		if (NorthPushed) {
			if ((yPole >= 0 && LastDecNorth))
				BumpSouth();
			else if (!LastDecNorth)
				BumpNorth();

			Ystate = StateVar::CorrectingN;

			i = (int)std::round(TrkRate / 4.0);

			CmdStr = "AY JG" + std::to_string(i) + ";";
			SendCommand(CmdStr);
		}

		if (SouthPushed) {
			if ((yPole >= 0 && !LastDecNorth))
				BumpNorth();
			else if (LastDecNorth)
				BumpSouth();

			Ystate = StateVar::CorrectingS;

			i = (int)std::round(TrkRate / 4.0);

			CmdStr = "AY JG-" + std::to_string(i) + ";";
			SendCommand(CmdStr);
		}
		break;

	case StateVar::CorrectingN:
	case StateVar::CorrectingS:
		if (!NorthPushed && !SouthPushed) {
			Ystate = StateVar::Tracking;
			SendCommand("AY ST;");
		}
		break;

	default:
		break;
	}
}